src/pemfile.rs

```diff
@@ -37,6 +43,56 @@ pub enum Item {
     Crl(CertificateRevocationListDer<'static>),
 }
 
+/// Errors that may arise when parsing the contents of a PEM file
+pub enum Error {
+    /// a section is missing its "END marker" line
+    MissingSectionEnd {
+        /// the expected "END marker" line that was not found
+        end_marker: Vec<u8>,
+    },
+
+    /// syntax error found in the line that starts a new section
+    IllegalSectionStart {
+        /// line that contains the syntax error
+        line: Vec<u8>,
+    },
+
+    /// base64 decode error
+    Base64Decode(DecodeError),
+}
+
+impl From<DecodeError> for Error {
+    fn from(v: DecodeError) -> Self {
+        Self::Base64Decode(v)
+    }
+}
+
+/// Extract and decode the next PEM section from `input`
+///
+/// - `Ok(None)` is returned if there is no PEM section to read from `input`
+/// - Syntax errors and decoding errors produce a `Err(...)`
+/// - Otherwise each decode section is returned with a `Ok(Some((Item::..., remainder)))` where
+///   `remainder` is the part of the `input` that follows the returned section
+pub fn read_one_from_slice(mut input: &[u8]) -> Result<Option<(Item, &[u8])>, Error> {
```

Comment:
I understand that this is trying to mirror the `read_one()` API and yet I wonder if that is actually good API -- for the conceptual action of "read one", should we treat what's represented here as `Ok(None)` as a kind of `Error` instead? And maybe we could attach this to `Item` as `Item::from_bytes()`?

Reply:
> should we treat what's represented here as Ok(None) as a kind of Error instead?

return something like an EOF error, say `Error::EmptySlice`? that seems reasonable at first but it may make the usage of the API a bit more clunky. right now you can write
``` rust
fn read_all_items(mut input: &[u8]) -> Result<Vec<Item>, Error> {
        let mut items = vec![];
        while let Some((item, rest)) = crate::read_one_from_slice(data)? {
            items.push(item);
            input = rest;
        }
        Ok(items)
}
```
because you can use the `?` operator on `read_from_one_from_slice`. if you were to return `Error::Eof` instead of `Ok(None)` you would have to do a more complex looking match statement to achieve the same
``` rust
fn read_all_items(input: &[u8]) -> Result<Vec<Item>, Error> {
    let mut items = vec![];
    loop {
        match crate::read_one_from_slice(data) {
            Err(Error::EmptySlice) => break Ok(items),
            Err(e) => return Err(e),
            Ok((item, rest)) => {
                items.push(item);
                input = rest;
            }
        }
    }
}
```

---

> And maybe we could attach this to Item as Item::from_bytes()?

from a readability stand point, I would expect (a) the signature of that to `fn(input: &[u8]) -> Result<Self, Error>` and (b) expect it to return an error if `input` contains more than one item or trailing bytes after one item.

I'm not opposed to having the function as an associated method on `Item` ( `fn(&[u8]) -> Result<(Self, &[u8]), Error>` does not quite fit the category of "constructor" in my mind )

Reply:
Yeah, maybe those ideas didn't quite make sense... Hmm.